Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1622 @@
+  map_mul := MonoidHom.map_mul'
+  map_one f := f.toOneHom.map_one'
+#align monoid_hom.monoid_hom_class MonoidHom.monoidHomClass
+#align add_monoid_hom.add_monoid_hom_class AddMonoidHom.addMonoidHomClass
+
+-- Porting note: help to_additive translate. This should ideally be handled by the tactic
+attribute [to_additive] MonoidHomClass.toOneHomClass
+
+/-- Any type that is a `MonoidHomClass` can be cast into a `MonoidHom`. -/
+@[to_additive "Any type that is a `AddMonoidHomClass` can be cast into a `AddMonoidHom`."]
+instance [MonoidHomClass F M N] : CoeTC F (M →* N) :=
+  ⟨fun f => { toFun := f, map_one' := map_one f, map_mul' := map_mul f }⟩
+/-- Any type that is a `AddMonoidHomClass` can be cast into a `AddMonoidHom`. -/
+instance [AddZeroClass M] [AddZeroClass N] [AddMonoidHomClass F M N] : CoeTC F (M →+ N) :=
+  ⟨fun f => { toFun := f, map_zero' := map_zero f, map_add' := map_add f }⟩
+
+@[simp, to_additive]
+theorem MonoidHom.coe_coe [MonoidHomClass F M N] (f : F) : ((f : M →* N) : M → N) = f := rfl
+#align monoid_hom.coe_coe MonoidHom.coe_coe
+#align add_monoid_hom.coe_coe AddMonoidHom.coe_coe
+
+@[to_additive]
+theorem map_mul_eq_one [MonoidHomClass F M N] (f : F) {a b : M} (h : a * b = 1) : f a * f b = 1 :=
+  by rw [← map_mul, h, map_one]
+#align map_mul_eq_one map_mul_eq_one
+#align map_add_eq_zero map_add_eq_zero
+
+@[to_additive]
+theorem map_div' [DivInvMonoid G] [DivInvMonoid H] [MonoidHomClass F G H]
+  (f : F) (hf : ∀ a, f a⁻¹ = (f a)⁻¹) (a b : G) : f (a / b) = f a / f b :=
+  by rw [div_eq_mul_inv, div_eq_mul_inv, map_mul, hf]
+#align map_div' map_div'
+#align map_sub' map_sub'
+
+/-- Group homomorphisms preserve inverse. -/
+@[simp, to_additive "Additive group homomorphisms preserve negation."]
+theorem map_inv [Group G] [DivisionMonoid H] [MonoidHomClass F G H]
+  (f : F) (a : G) : f a⁻¹ = (f a)⁻¹ :=
+  eq_inv_of_mul_eq_one_left <| map_mul_eq_one f <| inv_mul_self _
+#align map_inv map_inv
+#align map_neg map_neg
+
+/-- Group homomorphisms preserve division. -/
+@[to_additive "Additive group homomorphisms preserve subtraction."]
+theorem map_mul_inv [Group G] [DivisionMonoid H] [MonoidHomClass F G H] (f : F) (a b : G) :
+    f (a * b⁻¹) = f a * (f b)⁻¹ := by rw [map_mul, map_inv]
+#align map_mul_inv map_mul_inv
+#align map_add_neg map_add_neg
+
+/-- Group homomorphisms preserve division. -/
+@[simp, to_additive "Additive group homomorphisms preserve subtraction."]
+theorem map_div [Group G] [DivisionMonoid H] [MonoidHomClass F G H] (f : F) :
+  ∀ a b, f (a / b) = f a / f b := map_div' _ <| map_inv f
+#align map_div map_div
+#align map_sub map_sub
+
+-- Porting note: restore `to_additive`
+@[simp, to_additive]
+theorem map_pow [Monoid G] [Monoid H] [MonoidHomClass F G H] (f : F) (a : G) :
+  ∀ n : ℕ, f (a ^ n) = f a ^ n
+  | 0 => by rw [pow_zero, pow_zero, map_one]
+  | n + 1 => by rw [pow_succ, pow_succ, map_mul, map_pow f a n]
+#align map_pow map_pow
+#align map_smul map_smul
+
+attribute [to_additive_reorder 8] map_pow
```

Comment:
This is causing issues in the next declaration. 
This attribute should only be added if you manually make a `map_smul` lemma that has an argument order that differs from `map_pow`.

I'll try to add a feature of `to_additive` that will do the reordering of arguments itself when tagged with `[to_additive_reorder 8, to_additive]`